File: examples/python/rest/calendar_news_social.py
"""Copyright 2026 Volstrata.com

The `calendar`, `news`, `insider` and `social` domains: the list-shaped,
cursor-paged half of the surface.
Docs: https://volstrata.com/docs/api-catalog
      https://volstrata.com/docs/api-keys

    python rest/calendar_news_social.py

Eight capabilities, all GET, all collections:

    calendar.combined    GET /api/v1/calendar/combined    scheduled events, merged
    calendar.events      GET /api/v1/calendar/events      macro events
    calendar.corporate   GET /api/v1/calendar/corporate   company events
    news.feed            GET /api/v1/news/feed            headlines
    insider.filings      GET /api/v1/insider/filings      filed transactions
    social.sentiment     GET /api/v1/social/sentiment     per-ticker readings
    social.tallies       GET /api/v1/social/tallies       aggregate counts
    social.influencers   GET /api/v1/social/influencers   account-level rollup

Every one of them is at the Free floor in the catalog AND wants an authenticated
principal, so with no VOLSTRATA_API_KEY set each call answers

    401 {"code": "auth_required", ...}

and this file prints that and moves on. Any accepted credential clears it -- the
refusal is about having a principal, not about a plan. If you want to know which
capabilities your own credential can run, ask: GET /api/v1/meta/access, which
discover_capabilities.py prints.

Six of the eight share the cursor contract (`limit`, `cursor`, `next_cursor`)
with the rest of the paged surface, so once you can call one you can call those
with the same loop -- see pagination.py. `social.tallies` and
`social.influencers` are aggregates and take sample-size parameters instead of a
cursor, which is a good reason to read an operation's declared parameters in the
spec rather than assuming the family behaves alike.

Eight sequential requests, one page each.
"""

import os
import sys

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from volstrata_helpers import (  # noqa: E402  (import after the path fix, on purpose)
    attempt,
    get,
    heading,
    key_banner,
    row,
    run_examples,
    show,
)

# One page each. These run against live production and the point is the shape,
# not the volume.
PAGE_SIZE = 5


def _collection(payload):
    """Find the list in a paged response without hard-coding its key.

    Every paged envelope carries {ok, next_cursor, ...} plus exactly one
    collection, named for the thing it holds (events, items, filings, rows...).
    In your own code, look the name up once in the spec and use it directly;
    this generic version exists so one function can summarise eight endpoints.
    """
    best_key, best_rows = None, None
    for key, value in payload.items():
        if isinstance(value, list) and (best_rows is None or len(value) > len(best_rows)):
            best_key, best_rows = key, value
    return best_key, (best_rows or [])


ANSWERED = []


def _one_page(label: str, path: str, **params) -> None:
    """Fetch one page of a collection and summarise it, or report the refusal."""
    heading(label)
    payload = attempt(label, get, path, **params)
    if payload is None:
        return
    ANSWERED.append(label)
    key, rows = _collection(payload)
    row("collection", key)
    row("rows on this page", len(rows))
    row("next_cursor", payload.get("next_cursor") or "null (last page)")
    for counter in ("total", "count", "limit"):
        if counter in payload:
            row(counter, payload[counter])
    if rows:
        print("  First row:")
        show(rows[0], max_items=4, max_chars=700)


def calendars() -> None:
    """Three calendars, one shape.

    calendar.combined merges the other two, and takes `range` (7d, 30d, ...),
    `types`, `ticker`, `limit` and `cursor`. calendar.events adds
    `importance_min` and `region`; calendar.corporate is company-scoped and
    defaults to a 30-day range.

    Prefer `combined` unless you specifically need one kind: one request, one
    cursor, one merged ordering.
    """
    _one_page("calendar.combined", "/calendar/combined", range="7d", limit=PAGE_SIZE)
    _one_page("calendar.events", "/calendar/events", range="7d", limit=PAGE_SIZE)
    _one_page("calendar.corporate", "/calendar/corporate", range="30d", limit=PAGE_SIZE)


def news_and_filings() -> None:
    """news.feed and insider.filings: two collections, same cursor contract.

    news.feed takes an optional `ticker` -- omit it for everything, pass it to
    narrow. insider.filings takes `ticker` and `form` (default "4").

    Both are ordered newest first, so a client that polls should remember the
    newest item it has seen and stop reading when it reaches it, rather than
    re-reading a page every time. And poll on a timer you chose, not in a loop:
    rate limits are per owner and shared across every key you own.
    """
    _one_page("news.feed", "/news/feed", limit=PAGE_SIZE)
    _one_page("insider.filings", "/insider/filings", form="4", limit=PAGE_SIZE)


def social() -> None:
    """Three views of the same social data set, at three levels of aggregation.

    social.sentiment is per ticker and cursor-paged (`ticker`, `limit`,
    `cursor`). social.tallies is an aggregate (`top`, `min_posts`).
    social.influencers rolls up by account over a window (`hours`, `min`).

    `min_posts` and `min` are sample-size floors: raise them and thin rows drop
    out. A rate computed over three posts is not a rate, and the parameters are
    there so you can say what you consider enough.
    """
    _one_page("social.sentiment", "/social/sentiment", limit=PAGE_SIZE)
    _one_page("social.tallies", "/social/tallies", top=PAGE_SIZE)
    _one_page("social.influencers", "/social/influencers", hours=48)


def main() -> int:
    print("VolStrata: calendar, news and social -- https://volstrata.com")
    key_banner()
    exit_code = run_examples((
        ("calendars", calendars),
        ("news and filings", news_and_filings),
        ("social", social),
    ))
    if not ANSWERED:
        # Nothing answered, so the reader has seen eight refusals and no shape.
        # Here is the shape, written out by hand -- illustrative values, not a
        # captured response.
        heading("What a page looks like when it answers")
        print("""  {
    "ok": true,
    "next_cursor": "<opaque string, or null on the last page>",
    "limit": 5,
    "total": 128,
    "events": [
      {"id": "<opaque>", "ts": 1700000000, "ticker": "EXMPL",
       "title": "<one line>", "importance": 2, "kind": "<type>"}
    ]
  }""")
        print("  The keys that matter to your loop are `next_cursor` and the one")
        print("  collection. Everything else is domain detail. pagination.py runs")
        print("  the identical loop against a collection that needs no credential.")

    heading("Next")
    print("  pagination.py               the cursor loop, in full")
    print("  discover_capabilities.py    what your credential unlocks")
    print("  Keys: https://volstrata.com/api-keys")
    return exit_code


if __name__ == "__main__":
    sys.exit(main())

File: examples/javascript/rest/calendar_news_social.mjs
/**
 * Copyright 2026 Volstrata.com
 *
 * The `calendar.*`, `news.*`, `insider.*` and `social.*` domains — the
 * event-shaped half of the surface. Most of these are cursor-paged, so this is
 * where the `limit` and `cursor` parameters show up most often.
 * Catalog: https://volstrata.com/docs/api-catalog
 *
 * Run:  node rest/calendar_news_social.mjs [TICKER]
 *       npm run calendar
 *
 * Eight requests, each capped with a small `limit`. Only the first page of each
 * feed is fetched: pagination.mjs is where walking a feed to the end belongs,
 * and doing it eight times over in one script would spend a rate-limit budget
 * for no teaching benefit.
 *
 * Before you run this with no key: these capabilities currently answer
 * 401 `auth_required` to an anonymous caller, so a keyless run prints eight
 * refusals and no rows. That is a complete and correct run — just not an
 * interesting one. Read `rest/gex_levels.mjs` first if you have no key yet, and
 * come back to this file once you do. Any valid key gets past a 401; what your
 * plan then reaches is what `GET /api/v1/meta/access` reports.
 */

import {
  apiFetch,
  apiGet,
  banner,
  callCapability,
  heading,
  keysOf,
  report,
  show,
  showRateLimit,
  summarize,
} from "../lib/volstrata.mjs";

const ticker = (process.argv[2] ?? "SPX").toUpperCase();

banner(`calendar.*, news.*, insider.* and social.* — ${ticker}`);
show("ticker", `${ticker} (pass another as the first argument)`);

const results = [];

/**
 * Print the first page of a feed: how many rows, whether another page exists,
 * and a couple of rows so the shape is visible.
 *
 * `next_cursor` is the only termination condition that matters. A page that
 * comes back full is not proof there is more, and a short page is not proof
 * there is not.
 */
function showFeed(data, sampleCount = 3) {
  show("keys", keysOf(data));
  if ("total" in data) show("total", data.total);
  if ("limit" in data) show("limit applied", data.limit);
  if ("next_cursor" in data) {
    show("next_cursor", data.next_cursor === null ? "null (last page)" : "present (more pages)");
  }
  const [rowsKey, rows] = Object.entries(data).find(([, v]) => Array.isArray(v)) ?? [];
  if (rows) {
    show("rows key", rowsKey);
    show("rows", rows.length);
    for (const row of rows.slice(0, sampleCount)) {
      console.log(`      ${summarize(row, { depth: 1, maxKeys: 6, maxString: 90 })}`);
    }
  } else {
    show("payload", summarize(data, { depth: 2, maxKeys: 10 }));
  }
}

// ---------------------------------------------------------------------------
// calendar.combined — everything on the calendar, one call
// ---------------------------------------------------------------------------
//
// `range` takes a compact duration ("7d", "30d"), `types` filters the kinds of
// event, and `ticker` narrows to one symbol. Start with the combined feed and
// drop to the specific ones below when you need the extra filters they carry.

results.push(
  await callCapability("calendar.combined", async () => {
    const { res, data } = await apiFetch("/api/v1/calendar/combined", { range: "7d", limit: 5 });
    showFeed(data);
    showRateLimit(res);
    return data;
  }, { note: "range=7d, limit=5" }),
);

// ---------------------------------------------------------------------------
// calendar.events — macro events, with an importance filter
// ---------------------------------------------------------------------------

results.push(
  await callCapability("calendar.events", async () => {
    const data = await apiGet("/api/v1/calendar/events", {
      range: "7d",
      limit: 5,
      importance_min: 0, // raise this to drop the routine entries
      region: "all",
    });
    showFeed(data);
    return data;
  }, { note: "range=7d, importance_min=0" }),
);

// ---------------------------------------------------------------------------
// calendar.corporate — company events, over a longer default range
// ---------------------------------------------------------------------------

results.push(
  await callCapability("calendar.corporate", async () => {
    const data = await apiGet("/api/v1/calendar/corporate", { range: "30d", limit: 5 });
    showFeed(data);
    return data;
  }, { note: "range=30d, limit=5" }),
);

// ---------------------------------------------------------------------------
// news.feed — headlines, optionally for one symbol
// ---------------------------------------------------------------------------
//
// Leave `ticker` off for the whole feed. This repo does not name where any of
// this content originates; the fields are what the API returns.

results.push(
  await callCapability("news.feed", async () => {
    const data = await apiGet("/api/v1/news/feed", { ticker, limit: 5 });
    showFeed(data);
    return data;
  }, { note: `ticker=${ticker}, limit=5` }),
);

// ---------------------------------------------------------------------------
// insider.filings — regulatory filings
// ---------------------------------------------------------------------------
//
// `form` selects the filing type ("4" by default). Passing a `ticker` narrows
// to one issuer; leaving it off returns the general feed, which is what this
// call does so it returns rows for an index ticker too.

results.push(
  await callCapability("insider.filings", async () => {
    const data = await apiGet("/api/v1/insider/filings", { form: "4", limit: 5 });
    showFeed(data);
    return data;
  }, { note: "form=4, limit=5" }),
);

// ---------------------------------------------------------------------------
// social.sentiment, social.tallies, social.influencers
// ---------------------------------------------------------------------------
//
// Three different cuts of the same conversation: the per-post feed, the counted
// summary, and the accounts driving it. `tallies` and `influencers` take
// thresholds (`min_posts`, `min`) rather than a cursor — they are already
// aggregates.

results.push(
  await callCapability("social.sentiment", async () => {
    const data = await apiGet("/api/v1/social/sentiment", { ticker, limit: 5 });
    showFeed(data);
    return data;
  }, { note: `ticker=${ticker}, limit=5` }),
);

results.push(
  await callCapability("social.tallies", async () => {
    const data = await apiGet("/api/v1/social/tallies", { top: 10, min_posts: 3 });
    showFeed(data, 5);
    return data;
  }, { note: "top=10, min_posts=3" }),
);

results.push(
  await callCapability("social.influencers", async () => {
    const data = await apiGet("/api/v1/social/influencers", { hours: 48, min: 2 });
    showFeed(data, 5);
    return data;
  }, { note: "hours=48, min=2" }),
);

report(results);

heading("Related");
console.log("  Capability catalog   https://volstrata.com/docs/api-catalog");
console.log("  Pagination           node pagination.mjs");
console.log("  Next domain          node rest/accuracy_and_stats.mjs");
